### PR TITLE
Improve algorithm using related track similarity and add tests

### PR DESCRIPTION
**Pull request by lucyferrabee**

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -30,14 +21,43 @@ func main() {
...
+
+	return strings.TrimSpace(input), nil
+}
+
+func abs(n int) int {
+	if n < 0 {
+		return -n
+	}
+	return n
 }
```
Comment (jaddhx) [inline main.go]: Can't believe Go doesn't have Abs for integers builtin 😄 

You can write this as  `max(n, -n)`, which probably means you don't need the function and just do that inline at the callsites, though it's a question of style

File: auth/spotify_auth.go [hunk 2]
```diff
@@ -1,5 +1,3 @@
-// auth/spotify_auth.go
-
 package auth
```
Comment (jaddhx) [inline auth/spotify_auth.go:1]: I can't comment on the actual lines because it's not part of this PR, but I think the `Authenticator` struct is a bit weird.

I think it's better to just have a function return the `spotify.Client` directly, as the wrapping `Authenticator` struct isn't actually doing anything and has no methods on it.

File: main.go [hunk 3]
```diff
@@ -30,14 +21,43 @@ func main() {
...
 	auth := auth.NewAuthenticator("4779b9533e004287b6536fd8c5325adf", "8dbf0a481f8b4de0901fbc9661f6036c")
 
-	artistService := NewArtistService(auth)
 	relatedArtistService := NewRelatedArtistService(auth)
+	playlistService := NewPlaylistService(relatedArtistService)
 
-	artist := artistService.getByTrackId(id)
-	relatedArtists := relatedArtistService.getByArtistId(artist.id)
+	const targetPopularity = 70
+	const threshold = 2
```
Comment (jaddhx) [inline main.go]: Would be good to be able to configure these, rather than having them hard-coded

File: main.go [hunk 4]
```diff
@@ -30,14 +21,43 @@ func main() {
...
+	for _, track := range playlistTracks {
+		fmt.Printf("Name: %s, Artist: %s, Popularity: %d\n", track.Name, track.Artists[0].Name, track.Popularity)
+	}
+}
+
+type byPopularity []spotify.FullTrack
+
+func (a byPopularity) Len() int           { return len(a) }
+func (a byPopularity) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
+func (a byPopularity) Less(i, j int) bool { return a[i].Popularity > a[j].Popularity }
```
Comment (jaddhx) [inline main.go]: This shouldn't be in `main.go`, should be in `playlist.go` where you use it probably

Comment (jaddhx) [reply main.go]: Also I think the `Less` method is wrong, should be `return a[i].Popularity < a[j].Popularity`

File: main.go (hunk 1 above)
Comment (jaddhx) [reply main.go]: If you want to keep this method though I would create a `utils.go` file or something rather than having it in main

File: relatedartist_service_test.go [hunk 5]
```diff
@@ -0,0 +1,69 @@
+package main
```
Comment (jaddhx) [inline relatedartist_service_test.go:1]: Is this file ready? Seems unfinished / doesn't compile unless my setup is wrong.

File: relatedartist_service_test.go [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/mock"
+	"github.com/zmb3/spotify"
+)
+
+type RelatedArtistServiceTest struct {
+	Client SpotifyClient
+}
+
+func (s *RelatedArtistServiceTest) getFirstRelatedArtistByTrackId(id string) []relatedartist {
```
Comment (jaddhx) [inline relatedartist_service_test.go]: Bit confused about what this is meant to be doing, looks like you're writing test "service" code here and then your assertions later on only use this test service, rather than the actual service?

File: playlist.go [hunk 7]
```diff
@@ -1,37 +1,58 @@
...
+		if abs(track.Popularity-targetPopularity) <= threshold {
+			selectedTracks = append(selectedTracks, track)
+		}
+	}
+
+	return selectedTracks
+}
 
-	var playlist (string)
+func (ps *PlaylistService) removeDuplicates(tracks []spotify.FullTrack) []spotify.FullTrack {
```
Comment (jaddhx) [inline playlist.go:51]: Deduplicating an array is a pretty common thing, would be cool if this method worked for arrays of any type. Do you know about [generics](https://www.atatus.com/blog/generics-in-golang/)?

Comment (jaddhx) [reply playlist.go:51]: Also if you were to make a generic version of this, probably better to live in a `arrayutils` or something package, then you can reuse the same code elsewhere

File: main.go [hunk 8]
```diff
@@ -30,14 +21,43 @@ func main() {
...
-	relatedArtists := relatedArtistService.getByArtistId(artist.id)
+	const targetPopularity = 70
+	const threshold = 2
+
+	playlistTracks, err := playlistService.GeneratePlaylist(id, targetPopularity, threshold)
+	if err != nil {
+		log.Fatal(err)
+	}
 
 	fmt.Println("Here's your playlist: ")
```
Comment (jaddhx) [inline main.go]: What if the playlist is empty (e.g. because the threshold was too low)

File: playlist.go [hunk 9]
```diff
@@ -1,37 +1,58 @@
...
+func NewPlaylistService(relatedArtistService *RelatedArtistService) *PlaylistService {
+	return &PlaylistService{
+		RelatedArtistService: relatedArtistService,
+	}
+}
 
-	var tracks []spotify.FullTrack
+func (ps *PlaylistService) GeneratePlaylist(trackID string, targetPopularity, threshold int) ([]spotify.FullTrack, error) {
+	artist := ps.RelatedArtistService.getFirstRelatedArtistByTrackId(trackID)
+	allTopTracks, err := ps.RelatedArtistService.getTopTracksFromRelatedArtists(artist.id, 1)
```
Comment (jaddhx) [inline playlist.go:22]: Why only pass in `1` for the depth?

File: playlist.go [hunk 10]
```diff
@@ -1,37 +1,58 @@
...
+	uniqueTracks := ps.removeDuplicates(similarTracks)
 
-	return playlist{
-		tracks: tracks,
-	}
+	return uniqueTracks, nil
 }
 
-func displayPlaylist(p playlist) string {
+func (ps *PlaylistService) chooseSimilarPopularity(tracks []spotify.FullTrack, targetPopularity, threshold int) []spotify.FullTrack {
```
Comment (jaddhx) [inline playlist.go:35]: What's the thinking behind getting songs with similar popularities?

File: playlist_test.go [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+
+	result2 := ps.chooseSimilarPopularity(tracks, 90, 9)
+	assert.Len(t, result2, 1, "Expected 1 track")
+
+	result3 := ps.chooseSimilarPopularity(tracks, 60, 6)
+	assert.Len(t, result3, 2, "Expected 2 tracks")
+
+	for _, track := range result {
+		assert.True(t, abs(track.Popularity-70) <= 10, "Track popularity should be within the threshold")
+	}
```
Comment (jaddhx) [inline playlist_test.go]: Generally speaking, in unit tests you only want to do one set of assertions per test. That makes it easier to understand what the test is testing, and if there's a bug then you can immediately see which testcase is failing and why.

A better design would be to think about the edge cases, and to setup your test fixtures to check those edgecases specifically. E.g. you'd want to test:
 - All tracks being within the threshold
 - No tracks being within the threshold
 - One track being within the threshold
 - One track's popularity being exactly above the threshold
 - One track's popularity being exactly below the threshold
 - Zero threshold
 - Stupidly large threshold
 - Simliar things to do with the target popularity
 - Negative thresholds, negative popularities, empty track list, etc, anything else you can think of

Not sure how it works in go, but in other languages you can parameterise tests, so you just feed in the inputs and expected outputs to easily run the same test logic across multiple different inputs. If there's no support for parameterised tests in Go, then you can just write a helper function that does basically the same thing

Of course I've gone a bit overboard here for illustrative purposes, but it's good to think about all the edge cases and decide on the level of testing you think is appropriate

File: playlist_test.go [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+}
+
+func TestRemoveDuplicates(t *testing.T) {
+	ps := &PlaylistService{}
+
+	tracks := []spotify.FullTrack{
```
Comment (jaddhx) [inline playlist_test.go:147]: Fixture doesn't contain any duplicates so the test is a bit pointless. It's still a useful test for checking that nothing is removed when there are no duplicates of course, but the test name should reflect that

File: relatedartist.go [hunk 13]
```diff
@@ -24,29 +24,66 @@ func NewRelatedArtistService(auth *auth.Authenticator) *RelatedArtistService {
...
 		}
+		topTracks = append(topTracks, artistTopTracks...)
 
-		artists[i] = relatedartist{
-			name:       a.Name,
-			id:         string(a.ID),
-			popularity: a.Popularity,
-			topTracks:  topTracks,
+		if depth > 1 {
+			relatedArtistTopTracks, err := ras.getTopTracksFromRelatedArtists(artist.ID.String(), depth-1)
```
Comment (jaddhx) [inline relatedartist.go:50]: This recursive depth thing is interesting, but it could quickly blow up in terms of data volume. Space complexity here is something like `O(n!)` which is pretty scary :D.

The concept though of bringing in tracks from artists related to artists etc is pretty cool, but it'll need some contraints to be usable. E.g. max size of playlist, max number of songs per artists, etc

File: main.go (hunk 1 above)
Comment (lucyferrabee) [reply main.go]: i ended up using the "math" package which has Abs in it, although had to typecast the integers to float64, which isn't that pretty... but five minutes of googling would have lead me to find that Abs does exist, so that's my bad... :)

File: auth/spotify_auth.go (hunk 2 above)
Comment (lucyferrabee) [reply auth/spotify_auth.go:1]: yeah that makes sense

File: main.go (hunk 3 above)
Comment (lucyferrabee) [reply main.go]: i've moved them out into a config.json file

File: playlist.go (hunk 10 above)
Comment (lucyferrabee) [reply playlist.go:35]: i guess it's an arbitrary property that you can easily use to refine a search, it could be swapped in for another value held against a song if a good one existed... i guess the downside to it is that you then end up with a playlist full of songs with similar popularity, but the idea is you can then change how varied you go in this area, meaning if you want to generate a playlist with some similarly mainstream songs, and some niche songs, then you can widen the net. i would say it's not ideal though, and more an experiment in what results it produces than anything else.

File: playlist.go (hunk 9 above)
Comment (lucyferrabee) [reply playlist.go:22]: i think you answered your own question here further down on `relatedartist.go`, will answer more fully there...

File: relatedartist.go (hunk 13 above)
Comment (lucyferrabee) [reply relatedartist.go:50]: yeah, i quickly realised this when i put in 2 and realised it totally crashed haha. i guess for now it's a placeholder until I have a better way of handling related artists' related artists' related tracks, etc. it'll be built upon at a later date :D

File: playlist_test.go (hunk 12 above)
Comment (lucyferrabee) [reply playlist_test.go:147]: i agree

File: playlist.go [hunk 14]
```diff
@@ -35,8 +35,12 @@ func (ps *PlaylistService) GeneratePlaylist(trackID string, targetPopularity, th
 func (ps *PlaylistService) chooseSimilarPopularity(tracks []spotify.FullTrack, targetPopularity, threshold int) []spotify.FullTrack {
 	var selectedTracks []spotify.FullTrack
 
+	if len(tracks) == 0 {
+		return selectedTracks
+	}
+
```
Comment (jaddhx) [inline playlist.go:41]: Don't think you need this, if tracks == 0 then the for loop below won't have iterations so it'll return the empty `selectedTracks` array anyway

**Review by jaddhx — APPROVED** [review]
